### PR TITLE
import-boss reverse import

### PR DESCRIPTION
**Pull request by eloyekunle**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
This PR adds reverse import restrictions for packages by import-boss.

Reverse import restrictions help to define which packages in k/k can import certain other packages, defined inside the latter. This allows to have fine-grained import restrictions for "private packages" where we don't want to spread use inside of k/k.

The import-restrictions files added here specify that the `apiextensions-apiserver` package (and subpackages, except `v1` and `v1beta1`) should only be used in apiserver, and not anywhere else.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
Ref: https://github.com/kubernetes/gengo/pull/154

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```
/sig api-machinery

### CODE REVIEW

**Comment by caesarxuchao** [comment]
k8s/k8s already uses bazel, I think we can use the bazel visibility rules to replace import-boss, including the feature implemented by this PR. (Search "visibility" in [link](https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes)).

**Comment by sttts** [comment]
> k8s/k8s already uses bazel, I think we can use the bazel visibility rules to replace import-boss, including the feature implemented by this PR.

This can probably be done. But as we are using import-boss now, I very much prefer to add this small addition now. Whenever a bazel solution is available and it is more elegant, we can switch. It's nothing we expose outside of k/k, so a switch is always possible.

**Comment by caesarxuchao** [comment]
@sttts, that's fair. I'll leave this one to you.

/unassign
/assign @sttts

File: staging/src/k8s.io/code-generator/cmd/import-boss/main.go [hunk 1]
```diff
@@ -81,6 +81,30 @@ func main() {
...
+		"./vendor/k8s.io/kubectl/...",
+		"./vendor/k8s.io/kubelet/...",
+		"./vendor/k8s.io/kube-proxy/...",
+		"./vendor/k8s.io/kube-scheduler/...",
+		"./vendor/k8s.io/legacy-cloud-providers/...",
+		"./vendor/k8s.io/metrics/...",
+		"./vendor/k8s.io/node-api/...",
+		"./vendor/k8s.io/sample-apiserver/...",
+		"./vendor/k8s.io/sample-cli-plugin/...",
+		"./vendor/k8s.io/sample-controller/...",
```
Comment (sttts) [inline staging/src/k8s.io/code-generator/cmd/import-boss/main.go]: this list is misplaced here in the file. Even the three pre-existing packages above are wrong. We have to move the list where we invoke importboss (some hack/ script?).

And there iterate over `staging/src/k8s.io/*` and generate this list automatically.

**Comment by eloyekunle** [comment]
/remove-area conformance ipvs kubeadm kubelet release-eng
/remove-sig apps auth autoscaling contributor-experience network release scheduling storage testing

File: staging/src/k8s.io/code-generator/cmd/import-boss/main.go (hunk 1 above)
Comment (eloyekunle) [reply staging/src/k8s.io/code-generator/cmd/import-boss/main.go]: Done, now generating this list in [hack/verify-import-boss.sh](https://github.com/kubernetes/kubernetes/pull/83526/files#diff-d19bc83764e63f6805cab0f1ce53ea92R28)

File: staging/src/k8s.io/code-generator/go.mod [hunk 2]
```diff
@@ -17,7 +17,7 @@ require (
 	golang.org/x/net v0.0.0-20191004110552-13f9640d40b9 // indirect
 	golang.org/x/tools v0.0.0-20190920225731-5eefd052ad72 // indirect
 	gopkg.in/yaml.v2 v2.2.7 // indirect
-	k8s.io/gengo v0.0.0-20190822140433-26a664648505
```
Comment (sttts) [inline staging/src/k8s.io/code-generator/go.mod]: squash with 2nd commit?

File: go.sum [hunk 3]
```diff
@@ -590,8 +590,8 @@ gotest.tools/gotestsum v0.3.5/go.mod h1:Mnf3e5FUzXbkCfynWBGOwLssY7gTQgCHObK9tMpA
 grpc.go4.org v0.0.0-20170609214715-11d0a25b4919/go.mod h1:77eQGdRu53HpSqPFJFmuJdjuHRquDANNeA4x7B8WQ9o=
 honnef.co/go/tools v0.0.1-2019.2.2 h1:TEgegKbBqByGUb1Coo1pc2qIdf2xw6v0mYyLSYtyopE=
 honnef.co/go/tools v0.0.1-2019.2.2/go.mod h1:a3bituU0lyd329TUQxRnasdCoJDkEUEAqEt0JzvZhAg=
-k8s.io/gengo v0.0.0-20191010091904-7fa3014cb28f h1:eW/6wVuHNZgQJmFesyAxu0cvj0WAHHUuGaLbPcmNY3Q=
-k8s.io/gengo v0.0.0-20191010091904-7fa3014cb28f/go.mod h1:ezvh/TsK7cY6rbqRK0oQQ8IAqLxYwwyPxAX1Pzy0ii0=
```
Comment (sttts) [inline go.sum]: squash into bump commit

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/.import-restrictions [hunk 4]
```diff
@@ -1,20 +1,9 @@
...
-           "AllowedPrefixes": [
-           ],
-           "ForbiddenPrefixes": [
-             ""
-           ]
-       }
-   ]
-}
+inverseRules:
+  # Allow imports of this package in apiextensions-apiserver package and subpackages.
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/.import-restrictions]: try to explain why this is only allowed import. Something around the lines:

> Allow Internal packages only in apiextensions-apiserver itself, discourage use elsewhere.

File: pkg/printers/.import-restrictions [hunk 5]
```diff
@@ -1,10 +0,0 @@
-{
-  "Rules": [
-    {
-      "SelectorRegexp": "k8s[.]io/kubernetes/pkg/(api$|apis/)",
-      "ForbiddenPrefixes": [
-        "k8s.io/kubernetes/pkg/printers"
```
Comment (sttts) [inline pkg/printers/.import-restrictions]: why can this go?

Comment (eloyekunle) [reply pkg/printers/.import-restrictions]: There is a bug in the regular expressions in this import-restrictions file, so the forbidden package can never be selected by the `SelectorRegexp`.

If the intention was to forbid imports of `k8s.io/kubernetes/pkg/{api,apis}`, then I just updated the rule to reflect that.

Comment (sttts) [reply pkg/printers/.import-restrictions]: I think that was the case here: no internal package references from printers. @soltysh might know details.

File: pkg/cloudprovider/providers/.import-restrictions [hunk 6]
```diff
@@ -3,14 +3,7 @@
...
 			"AllowedPrefixes": [
-				"k8s.io/utils/exec",
-				"k8s.io/utils/io",
-				"k8s.io/utils/keymutex",
-				"k8s.io/utils/mount",
-				"k8s.io/utils/net",
-				"k8s.io/utils/nsenter",
-				"k8s.io/utils/path",
-				"k8s.io/utils/strings"
+				""
```
Comment (liggitt) [inline pkg/cloudprovider/providers/.import-restrictions]: doesn't this change allow anything? if so, why do we still need this stanza for k8s.io/utils?

Comment (eloyekunle) [reply pkg/cloudprovider/providers/.import-restrictions]: okay, we actually don't need the stanza here

**Comment by eloyekunle** [comment]
ping @liggitt

File: pkg/controller/.import-restrictions [hunk 7]
```diff
@@ -11,309 +11,6 @@
...
-        "golang.org/x/net/context",
-        "google.golang.org/api/compute/v1",
-        "google.golang.org/api/googleapi",
-        "google.golang.org/api/compute/v0.alpha",
-        "google.golang.org/api/container/v1",
-        "google.golang.org/api/compute/v0.beta",
-        "google.golang.org/api/tpu/v1",
-        "golang.org/x/net/context",
-        "google.golang.org/grpc"
-      ]
```
Comment (sttts) [inline pkg/controller/.import-restrictions]: these removal look good to me. This is not the intended use of import-boss.

File: pkg/cloudprovider/providers/.import-restrictions [hunk 8]
```diff
@@ -1,18 +1,5 @@
...
-				"k8s.io/utils/io",
-				"k8s.io/utils/keymutex",
-				"k8s.io/utils/mount",
-				"k8s.io/utils/net",
-				"k8s.io/utils/nsenter",
-				"k8s.io/utils/path",
-				"k8s.io/utils/strings"
-			]
-		},
 		{
```
Comment (sttts) [inline pkg/cloudprovider/providers/.import-restrictions]: lgtm. k8s.io/utils is not a staging repo and we should not restrict use.

File: pkg/master/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go [hunk 9]
```diff
@@ -109,7 +108,7 @@ func NewClusterAuthenticationTrustController(requiredAuthenticationData ClusterA
 				return cast.Name == configMapName
 			}
 			if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
-				if cast, ok := tombstone.Obj.(*apiextensions.CustomResourceDefinition); ok {
+				if cast, ok := tombstone.Obj.(*corev1.ConfigMap); ok {
```
Comment (liggitt) [inline pkg/master/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go]: that's exciting

**Review by liggitt — COMMENTED** [review]
/lgtm
/approve

File: pkg/master/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go (hunk 9 above)
Comment (eloyekunle) [reply pkg/master/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go]: Very exciting :grin:

**Pull request merged**